src/kademlia/node.cpp

```diff
@@ -1210,111 +1019,65 @@ void node::incoming_request(msg const& m, entry& e)
 			VALGRIND_CHECK_MEM_IS_DEFINED(msg_keys[4].string_ptr(), item_sig_len);
 			VALGRIND_CHECK_MEM_IS_DEFINED(pk, item_pk_len);
 #endif
+			boost::int64_t seq = msg_keys[2].int_value();
+
+			if (seq < 0)
+			{
+				m_counters.inc_stats_counter(counters::dht_invalid_put);
+				incoming_error(e, "invalid (negative) sequence number");
+				return;
+			}
+
 			// msg_keys[4] is the signature, msg_keys[3] is the public key
 			if (!verify_mutable_item(buf, salt
-				, msg_keys[2].int_value(), pk, sig))
+				, seq, pk, sig))
 			{
 				m_counters.inc_stats_counter(counters::dht_invalid_put);
 				incoming_error(e, "invalid signature", 206);
 				return;
 			}
 
-			dht_mutable_table_t::iterator i = m_mutable_table.find(target);
-			if (i == m_mutable_table.end())
-			{
-				// this is the case where we don't have an item in this slot
-				// make sure we don't add too many items
-				if (int(m_mutable_table.size()) >= m_settings.max_dht_items)
-				{
-					// delete the least important one (i.e. the one
-					// the fewest peers are announcing)
-					dht_mutable_table_t::iterator j = std::min_element(m_mutable_table.begin()
-						, m_mutable_table.end()
-						, boost::bind(&dht_immutable_item::num_announcers
-							, boost::bind(&dht_mutable_table_t::value_type::second, _1)));
-					TORRENT_ASSERT(j != m_mutable_table.end());
-					free(j->second.value);
-					free(j->second.salt);
-					m_mutable_table.erase(j);
-					m_counters.inc_stats_counter(counters::dht_mutable_data, -1);
-				}
-				dht_mutable_item to_add;
-				to_add.value = static_cast<char*>(malloc(buf.second));
-				to_add.size = buf.second;
-				to_add.seq = msg_keys[2].int_value();
-				to_add.salt = NULL;
-				to_add.salt_size = 0;
-				if (salt.second > 0)
-				{
-					to_add.salt = static_cast<char*>(malloc(salt.second));
-					to_add.salt_size = salt.second;
-					memcpy(to_add.salt, salt.first, salt.second);
-				}
-				memcpy(to_add.sig, sig, sizeof(to_add.sig));
-				TORRENT_ASSERT(sizeof(to_add.sig) == msg_keys[4].string_length());
-				memcpy(to_add.value, buf.first, buf.second);
-				memcpy(&to_add.key, pk, sizeof(to_add.key));
-		
-				boost::tie(i, boost::tuples::ignore) = m_mutable_table.insert(
-					std::make_pair(target, to_add));
-				m_counters.inc_stats_counter(counters::dht_mutable_data);
+			TORRENT_ASSERT(item_sig_len == msg_keys[4].string_length());
 
-//				fprintf(stderr, "added mutable item (%d)\n", int(m_mutable_table.size()));
+			boost::int64_t item_seq;
+			if (!m_storage->get_mutable_item_seq(target, item_seq))
```

Comment:
I suppose this is one of those cases of abstractions taking a toll in performance. However, and only to comment about it, the double lookup is only evident if we know how the current default storage is implemented. The semantic of `get_mutable_item_seq` and `put_mutable_item` taken together does not imply the need of the double lookup.